content/doc/book/pipeline/reference.adoc

```diff
@@ -0,0 +1,492 @@
+closures.
+* _Parameters_: Either a `Closure` of one key and either a single value
+or a `Closure` of multiple configuration pairs, or one of two constants
+- `none` or `any`.
+** _Map Keys_:
+*** Note that this will be an `ExtensionPoint`, so plugins will be able
+to add to the available image providers.
+*** `docker`
+**** _Description_: If given, uses this Docker image for the container
+the build will run in. If no `label` is given, the container will be run
+within a simple `node { ... }` block.
+*** `dockerfile`
+**** _Description_: If given, builds from the Dockerfile in the
+repository and runs in a container using the resulting image.
+*** `label`
+**** _Description_: If given, uses this label for the node the build
+will run on - if `docker` is also specified, the container will run on
+that node.
+** `none`
+*** _Type_: bareword
+*** _Description_: If given, node/image management will need to be
+specified explicitly in stages and no automatic `checkout scm` call will
+occur.
+** `any`
+*** _Type_: bareword
+*** _Description_: If given, any available agent will be used.
+* _Takes a Closure_: yes
+* _Examples_:
+
+[pipeline]
+----
+// Declarative //
+agent {
+    label 'hi-speed'
+}
+----
+[pipeline]
+----
+agent {
+    docker 'ubuntu:lts'
+}
+----
+[pipeline]
+----
+// Declarative //
+agent {
+    docker {
+        image 'ubuntu:lts'
+        label 'has-docker'
+        args "-v /tmp:/tmp -p 80:80"
+    }
+}
+----
+[pipeline]
+----
+// Declarative //
+agent none
+----
+[pipeline]
+----
+// Declarative //
+agent any
+----
+
+[[environment]]
+=== environment
+
+* _Description_: A sequence of `key = value` pairs, which will be passed
+to the `withEnv` step the build will be executed within.
+* _Required_: No
+* _Allowed In_: Top-level `pipeline` or `stage` closures only.
+* _Parameters_: None
+* _Takes a Closure_: Yes
+* _Closure Contents_: One or more lines with `foo = 'bar'` variable
+name/value pairs.
+** The name doesn't need to be quoted.
+** Bind credentials using the "credentials('<id>')" function. _NOTE_:
+credentials binding requires that the pipeline or stage is running
+within an agent. In the example below the credentials gets bound in
+slightly different ways depending of the type of the key;
+*** If it is a `Secret Text` the variable `SAUCE_ACCESS` will contain
+that text.
+*** If it is a `Secret File` the variable `SAUCE_ACCESS` will contain a
+path to the file on the build agent.
+*** If it is a `Standard username and password` credential, three
+variables will be added to the environment:
+**** `SAUCE_ACCESS` containing '<username>:<password>'
+**** `SAUCE_ACCESS_USR` containing the username
+**** `SAUCE_ACCESS_PSW` containing the password
+* _Examples_:
+
+[pipeline]
+----
+// Declarative //
+environment {
+   CXX          = 'g++-4.8'
+   SAUCE_ACCESS = credentials('sauce-lab-dev')
+   someVar      = 'someValue'
+}
+----
+
+[[stages]]
+=== stages
+
+* _Description_: A sequence of one or more Pipeline `stage`, each of
```

Comment:
line 143: "one or more" has to accommodate plural, so it should say "Pipeline stages"